The GHC20xx process  (under review) (#372)
ghc-proposals/ghc-proposals: proposals/0000-ghc-extensions.rst

The GHC20XX process
===================

.. author:: Alejandro Serrano, Joachim Breitner
.. date-accepted:: 
.. proposal-number::
.. ticket-url::
.. implemented::
.. highlight:: haskell
.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/372>`_.
.. sectnum::
.. contents::

> This proposal sets up a process to introduce new language versions `GHC20XX`, extending `Haskell2010` as defined by the Report with several extensions that are considerd harmless, helpfun and uncontroversial.

Reviewer: ```suggestion
This proposal sets up a process to introduce new language versions `GHC20XX`, extending `Haskell2010` as defined by the Report with several extensions that are considerd harmless, helpful and uncontroversial.
```

Author: Ah, no fun?

Motivation
----------

It has almost become a meme that any slightly-complex Haskell module starts with a dozen lines of ``{-# LANGUAGE #-}`` pragmas. This poses several problems:

- beginners often see features behind an extension as "advanced" or "obscure", even though some like ``LambdaCase`` or ``NamedFieldPuns`` are quite simple;
- some sorts of programming in Haskell – mostly type-level oriented – require enables lots of extensions (think of the usual combo ``MultiParamTypeClasses`` + ``FlexibleInstances`` + ``FlexibleContexts``). Sometimes the fact that there are different extensions, instead of a single `ExtendedTypeClasses` is mostly historical.

> For that reason, it's benefitial to pack those extensions which have "graduated" in terms of stability and usefulness into a single language version.

Reviewer: ```suggestion
For that reason, it's beneficial to pack those extensions which have "graduated" in terms of stability and usefulness into a single language version.
```

FWIW I think this idea of a feature “graduating” is an _excellent_ litmus test.

> We aim to lean on the conservative side in our selection of extensions: Not including an extension that we should have is relatively harmless (users of that extension will have to still set that manually, but they alreay have to do that, and we can add it alter), while including an extension that we should not have hurts more (any detriminal effect on developer experience now affects many users).

Reviewer: ```suggestion
We aim to lean on the conservative side in our selection of extensions: Not including an extension that we should have is relatively harmless (users of that extension will have to still set that manually, but they alreay have to do that, and we can add it later), while including an extension that we should not have hurts more (any detriminal effect on developer experience now affects many users).
```

Reviewer: This is a bit of a run-on sentence, and I found it a bit hard to parse. Maybe breaking it up after the first parenthetical:

> …we can add it later). Conversely, including an extension which we shouldn’t have hurts more…

Author: ```suggestion
We aim to lean on the conservative side in our selection of extensions: Not including an extension that we should have is relatively harmless (users of that extension will have to still set that manually, but they already have to do that, and we can add it alter). Conversely, including an extension that we should not have hurts more (any detriminal effect on developer experience now affects many users).
```

> We also aim to have a process that is efficient. For each extension, we are all very capable of having very long discussions about the merits, up-sides and downsides, and almost anybody will have an opinion on almost any extensions. But aiming to include only _uncontroversial_ extensions, we aim to avoid much of these discussion by construction. The process deliberately does not provide room for long per-extensions deliberations, neither within the committee or with the overall community, while still involving the community.

Reviewer: I’m curious what “uncontroversial” means, but haven’t finished reading yet. I don’t know if that suggests the doc should discuss this more up front or that it’s working well as a hook, tho 😄

Author: Hmm, is it unclear? My intention is that it means “there is no (or little) controversy about turning this on by default”. Maybe it gets clearer later… if not, please do suggest improvements to the wording!

Proposed Change Specification
-----------------------------

> We propose to introduce a process to develop new "language versions", dubbed ``GHC20XX``, which extend ``Haskell2010`` with some default extensions. This proposal does not set *which* extensions are included in the first version,, rather seeks to provide a framework to reach a conclusion in a constructive and timely manner.

Reviewer: ```suggestion
We propose to introduce a process to develop new "language versions", dubbed ``GHC20XX``, which extend ``Haskell2010`` with some default extensions. This proposal does not set *which* extensions are included in the first version, but rather seeks to provide a framework to reach a conclusion in a constructive and timely manner.
```

Criteria
^^^^^^^^

We define a set of criteria that an extension should guide the committee members in their process of picking suitable extensions. These are not hard criteria, but often subjective, and not all need to be satisfied in all cases.

1. The extension is (mostly) *conservative*: All programs that were accepted without the extension remain accepted, and with the same meaning. We prefer to take this as a strict criteria, and grant exceptions on a case-by-case basis.

> 2. *New failure modes* that become possible with the extension are *rare* and/or easy to diagnose. These failure modes include new error messages, wrong inferred types, and runtime errors, for example. In other words, the *developer experience* should not get worse by adding this extension to the default mix.

Reviewer: Is it worth giving an example? e.g. something that changes the semantics of existing programs (e.g. how `-XScopedTypeVariables` can cause type variables to start unifying in e.g. a `where` clause in an `instance` declaration)

Reviewer 2: Maybe `ScopedTypeVariables` is not the best example, because I guess it might be one of the extensions for which some of the requirements are waived due to its overwhelming use.

3. The extensions *complement* the design of standard Haskell. For example, ``MultiParamTypeClasses`` extends what's already there, and works well with the rest of features. If we suddenly decided you can use Lisp-like syntax, that would not complement the current design.
4. The extension has been – and can reasonably be predicted to remain – *stable*.
5. The extension is one that users might plausibly want to be on all the time. This excludes experimental extensions that deliberately enable potentially-unsafe or unstable features, such as ``IncoherentInstances`` or ``MagicHash``.

> 6. The extension has *widespread* usage.

Reviewer: …as measured by code on e.g. hackage, presumably?

Reviewer 2: Yes, and also about what people say in the polls.

> 5. The extension is *not* intentional opt-in as a *gate-keeper* of advanced or potentially-unsafe feature. ``IncoherentInstances`` should not be enabled by default because of its potential unsafety, and the point of ``MagicHash`` is precisely that it it unlocks names not there by default.

Author: ```suggestion
5. The extension is one that users might plausibly want to be on all the time. This excludes experimental extensions that deliberately enable potentially-unsafe or unstable features, such as ``IncoherentInstances`` or ``MagicHash``.
```

7. The extension is favored by the community, with many in favor, and very few opposed to its inclusion.

Process
^^^^^^^

* 4 months before the expected GHC spring release day of 202x, the committee Secretary starts the GHC202x process.

>   They inform the committe, in an email to the mailing list, of all language extensions supported by the latest released GHC that are not in GHC202(x-1), which could be added. They also list all extensions *in* GHC202(x-1), which could be removed.

Reviewer: ```suggestion
  They inform the committee, in an email to the mailing list, of all language extensions supported by the latest released GHC that are not in GHC202(x-1), which could be added. They also list all extensions *in* GHC202(x-1), which could be removed.
```

Removed from 202x, right? Not retroactively dropped from 202(x-1)?

Author: ```suggestion
  They inform the committee, in an email to the mailing list, of all language extensions supported by the latest released GHC that are not in GHC202(x-1), which could be added. They also list all extensions *in* GHC202(x-1), which might be omitted in GHC202(x-1) (likely a rare thing).
```

  The secretary also creates a PR with a proposal saying (roughly) 
   
    GHC202x contains the following extensions in addition to those in GHC202(x-1):
    
    * (none yet)
    
    and removes these extensions
   
    * (none yet)

* In order to gather data on the criterium “widespread usage”, the secretary creates a tally of which extensions are used how often on hackage.

> * In order to gather data on the criterium “community spport”, the secretary runs a public poll on a suitable platform for one week where anyone

Reviewer: ```suggestion
* In order to gather data on the criterium “community support”, the secretary runs a public poll on a suitable platform for one week where anyone
```

>   can vote in favor or against the inclusion of a given extension.  

Reviewer 3: It's not exactly the same, but I have asked "Which language extensions would you like to be enabled by default?" in the state of Haskell survey for the past three years:

- https://taylor.fausak.me/2019/11/16/haskell-survey-results/#s2q5
- https://taylor.fausak.me/2018/11/18/2018-state-of-haskell-survey-results/#question-038
- https://taylor.fausak.me/2017/11/15/2017-state-of-haskell-survey-results/#question-21

I plan to ask it again this year. If there's something I can do to make it more useful to this GHC20xx process, please let me know!

Author: @simonmar actually just mentioned that, and maybe that is simply enough to get some community data. We could make the process flexible in this point:
```suggestion
* In order to gather data on the criterium “community support”, the secretary runs a public poll on a suitable platform for one week where anyone
  can vote in favor or against the inclusion of a given extension, or points the committee to a suitable existing survey result.
```

Author: If I’d do a dedicated poll, I might ask for “Yes please”/“Don’t care”/“No, please don’t” to also gauge how much resistance it might have. But maybe not worth the bother.

Reviewer 3: I like how that sounds! Other people have asked for something like this before. I've held off because it's already an intimidating question. But I think I could create a simple enough UI/UX if "Yes please" is 👍/upvote, "No, please don't" is 👎/downvote, and "Don't care" is no vote at all.

* Within two weeks of the start of the process, every committee member is expected to send an initial list of which extensions they expect to be in GHC202x to the mailing list.
   

>   Committe members are expected to take the Hackage statistics and the community vote into account.

Reviewer: ```suggestion
  Committee members are expected to take the Hackage statistics and the community vote into account.
```

>   These mails may contain justifications for why a certain extension is or is not included, but not required (or even expected).

Reviewer: ```suggestion
  These mails may contain justifications for why a certain extension is or is not included, but this is not required (or even expected).
```

   
  Long discussions of individual extensions are discouraged at this point. If there is controversy around an extension, it is a strong sign that it should simply not be included.

* After these four weeks, the proposal with the current tally gets accepted by the secretary, and defines GHC202x

Cadence
^^^^^^^

Likely, the first iteration of this process will be vastly different from the following ones: The first one is expected to add a large number of uncontroversial extensions; so the next iteration will likely only make a smaller, but more controversial change.

Therefore, this proposal does *not* commit to a fixed cadence. Instead, 6 months after the first release of a version of GHC that supports a GHC20XX set, we evaluate the outcome, the process, and the perceived need of a next release. At that time we will refine the processes, if needed, and set a cadence.

Costs and Drawbacks
-------------------

The implementation cost seems small.

The cost of a GHC20xx extension is that, upon reading a file with ``{-# LANGUAGE GHC20xx #-}``, the reader does not immediatelly know the set of enabled extensions; this may hamper readability of code.

> The costs of this process is that it binds volunteer time, and there is is risk of unpleasant, heated debates, because everybody has opinions. The process tries to err on the conversative side and rather add too few than too many extensions.

Reviewer: ```suggestion
The costs of this process is that it binds volunteer time, and there is is risk of unpleasant, heated debates, because everybody has opinions. The process tries to err on the conservative side and rather add too few than too many extensions.
```

   opposed to its inclusion.

Process
^^^^^^^

* 4 months before the expected GHC spring release day of 20xx, the committee
  Secretary starts the GHC20xx process.

  They inform the committee, in an email to the mailing list, of all language
  extensions supported by the latest released GHC that are not in GHC20(xx-1),
  which could be added. They also list all extensions *in* GHC20(xx-1), which
  might be omitted in GHC202(xx-1) (likely a rare thing).

* In order to gather data on the criterium “widespread usage”, the secretary
  creates a tally of which extensions are used how often on Hackage.

* In order to gather data on the criterium “community support”, the secretary
  runs a public poll on a suitable platform for one week where anyone can vote
  in favor or against the inclusion of a given extension, or points the
  committee to a suitable existing survey result.

**Alternative 1:**

* At the start of the process, the secretary creates a PR with a proposal saying (roughly)
   
    GHC20xx contains the following extensions in addition to those in
    GHC20(xx-1):
    
    * (none yet)
    
    and removes these extensions
   
    * (none yet)

> 

Author: ```suggestion

    This PR is a suitable place, besides the poll, for the wider commity to weigh in.
```
@simonpj  like this?

Author: ```suggestion

    This PR is a suitable place, besides the poll, for the wider
     community to weigh in. The community is invited to follow the
     committee votes, rationales and discussion on the public email
     archive, and if the committee is missing an important piece of
     information (e.g. more code breaking than expected), to raise
     such a point.
 
     We hope, however, that the community poll is sufficient to convey
     the level of community support and demand for specific extensions
     have, and want to discourage lengthy, opinion-based discussions of
     the merits of extensions.
 
     If you miss your favorite extension in the list, please remember
     that you can still use it (by setting the flag explicitly), and
     that it can still go in next round.
```
